src/muse/constraints.py

```diff
@@ -493,6 +494,64 @@ def max_production(
     )
 
 
+@register_constraints
+def demand_limitting_capacity(
+    demand_: xr.DataArray,
+    assets: xr.Dataset,
+    search_space: xr.DataArray,
+    market: xr.Dataset,
+    technologies: xr.Dataset,
+    year: int | None = None,
+) -> Constraint:
+    """Limits the maximum combined capacity to match the demand.
+
+    This is a somewhat more restrictive constraint than the max_production constraint or
+    the maximum capacity expansion. In this case, the combined new capacity of all
+    assets must be sufficient to meet the demand of the most demanding timeslice, and
+    no more.
+
+    Rather than coding from scratch the constraint, we can use the max_production
+    constraint and the demand constraint to construct this constraint. Starting from
+    the maximum production instead of the maximum capacity ensures that the constraint
+    accounts for the utilization factor of the technologies.
+    """
+    # We start with the maximum production constraint and the demand constraint
+    capacity_constraint = max_production(
+        demand_, assets, search_space, market, technologies, year=year
+    )
+    demand_constraint = demand(
+        demand_, assets, search_space, market, technologies, year=year
+    )
+
+    # We need to find the most demanding timeslice and use as the demand constraint
+    b = (
+        demand_constraint.b.max("timeslice")
+        if "timeslice" in demand_constraint.b.dims
+        else demand_constraint.b
+    )
+
+    # Now we need to find the maximum capacity constraint (as the capacity here is
+    # negative), and switch the sign to make it positive. In practice, we want to
+    # capture the utilisation factor of the technologies.
+    capacity = (
+        -capacity_constraint.capacity.max("timeslice")
```

Comment:
Why the `max` operation? I guess capacity will be the same in all timeslices so this is just to reduce the data over the timeslice dimension? Unless I'm misunderstanding something

Reply:
The capacity is not really "capacity", but the capacity-related component of the constraint, which is a multiplying factor to the actual capacity being installed. The capacity itself is, naturally, the same for all timeslices but this factor also includes the utilisation factor, which can be timeslice dependent. By taking the "max" we make sure that the timeslice with the minimum utilisation factor is covered.